Update Tuya Integration with Tuya v2 (#19519)
home-assistant/home-assistant.github.io: source/_integrations/tuya.markdown

title: Tuya
description: Instructions on how to setup the Tuya hub within Home Assistant.
ha_category:
  - Climate
  - Cover
  - Fan
  - Light
  - Scene
  - Switch
ha_iot_class: Cloud Polling

> ha_release: 2021.9.7

Reviewer: The release is still 0.74 when it was first introduced, no need to update this.

Author: fixed.

ha_config_flow: true

> ha_domain: Tuya

Reviewer: All domains are lowercase. It's the directory name in HA.

```suggestion
ha_domain: tuya
```

Author: fixed.

ha_dhcp: true
---

> The Tuya Home Assistant integration is developed for controlling **Powered by Tuya (PBT)** devices using the [tuya-iot-python-sdk](https://github.com/tuya/tuya-iot-python-sdk) (a Python version of [Tuya Open API](https://developer.tuya.com/en/docs/iot/api-reference?id=Ka7qb7vhber64)), which is officially maintained by the Tuya Developer Team.

Reviewer: Current sentence only mentions the SDK is official. I would leave out the SDK and emphasize that this is an official one.

```suggestion
The Tuya Home Assistant integration allows you to control any **Powered by Tuya** devices that you have added to the Tuya Smart or Smart Life apps. This integration is official and maintained by Tuya.
```

Author: fixed

> It applies to smart devices that you have added to the Tuya Smart or Smart Life app.

Reviewer: ```suggestion
{% include integrations/config_flow.md %}
```

Author: fixed.

## Supported device categories
- **Climate**: Supports air conditioner, heater, thermostat.
- **Cover**: Supports curtain, curtain switch.
- **Fan**: Supports fan, air purifier, humidifier, dehumidifier.
- **Light**: Supports light, light strip, ambient light, light string, humidifier's light, Ceiling light.
- **Scene**: Supports smart scene.
- **Switch**: Supports switch, socket, power strip, smart kettle, breaker, pet water feeder, air purifier, diffuser.

## Quick installation and startup

### Configure on the Tuya IoT Platform

#### Prerequisites

You have created an account for [Tuya Smart or Smart Life app](https://developer.tuya.com/en/docs/iot/tuya-smart-app-smart-life-app-advantages?id=K989rqa49rluq#title-1-Download).

#### Create a project

1. Log in to the [Tuya IoT Platform](https://iot.tuya.com/).
2. In the left navigation bar, click **Cloud** > **Projects**.
3. On the page that appears, click **Create**.
4. In the **Create Project** dialog box, configure **Project Name**, **Description**, **Industry**, and **Availability Zone**. In the **Development Method** field, select **Smart Home PaaS** from the dropdown list.
	> **Note:** Select all available zones.

> 	![](https://airtake-public-data-1254153901.cos.ap-shanghai.myqcloud.com/content-platform/hestia/1622030427d285b460b84.png)

Reviewer: All images need to be included in the PR and hosted on home-assistant.io. No external URLs for images.

Author: replace with local images in images folder.

![](https://airtake-public-data-1254153901.cos.ap-shanghai.myqcloud.com/content-platform/hestia/1627013411144507956ff.png)

#### Enter your Tuya credential

In the Tuya Integration window, select **Smart Home PaaS** and click **SUBMIT**.

![](https://airtake-public-data-1254153901.cos.ap-shanghai.myqcloud.com/content-platform/hestia/1627379701127540ff6a2.png)

See the following table and enter your Tuya credential.

| Field | Description |
| ------- | -------- |
| Access ID and Access Secret| Go to your cloud project on [Tuya IoT Platform](https://iot.tuya.com/cloud/?_source=3a6f90da0e85f686f89c4f85c883e8f0). Find the **Access ID** and **Access Secret** on the **Project Overview** tab.|
| Mobile App | Must select the one you used to link devices on the Tuya IoT Platform. |
| Country Code | The country you select on logging in to Tuya Smart or Smart Life app.|
| Account | Tuya Smart or Smart Life app account. |
| Password | The password of your app account. |
> **Note**: The app mentioned in the table must be the one you used to link devices on the Tuya IoT Platform.
</details>

> ## Errorcode and troubleshooting

Reviewer: ```suggestion
## Error codes and troubleshooting
```

Author: fixed

## Related Documents

### How to Contribute
* [How to Develop a New Driver](https://github.com/tuya/tuya-home-assistant/wiki/How-to-Develop-a-New-Driver)
* [How to Get the Log](https://github.com/tuya/tuya-home-assistant/wiki/How-to-Get-the-Log)
* [Supported Device Categories](https://github.com/tuya/tuya-home-assistant/wiki/Supported-Device-Category)

### Documentation
* [How does the Integration work](https://github.com/tuya/tuya-home-assistant/wiki/How-the-Integration-works)
* [Set up Home Assistant on a Raspberry Pi](https://github.com/tuya/tuya-home-assistant/wiki/Set-up-Home-Assistant-on-a-Raspberry-Pi)
* [FAQs](https://github.com/tuya/tuya-home-assistant/wiki/FAQs)

> 
> 
> {% include integrations/config_flow.md %}

Reviewer: ```suggestion
```

Author: fixed.

ha_dhcp: true
---

The Tuya Home Assistant integration is developed for controlling **Powered by Tuya (PBT)** devices, which is officially maintained by the Tuya Developer Team.

You can find more information for this integration on [Tuya Integration Github](https://github.com/tuya/tuya-home-assistant?_source=418498f8d2819b5b1654a0c12c5325cd).

It applies to smart devices that you have added to the Tuya Smart or Smart Life app.

{% include integrations/config_flow.md %}

> ## Supported device categories

Reviewer 2: We call them platforms.

```suggestion
## Supported platforms

```

Author: fixed.

- **Climate**: Supports air conditioner, heater, thermostat.
- **Cover**: Supports curtain, curtain switch.
- **Fan**: Supports fan, air purifier, humidifier, dehumidifier.
- **Light**: Supports light, light strip, ambient light, light string, humidifier's light, Ceiling light.
- **Scene**: Supports smart scene.
- **Switch**: Supports switch, socket, power strip, smart kettle, breaker, pet water feeder, air purifier, diffuser.

## Quick installation and startup

### Configure on the Tuya IoT Platform

> #### Prerequisites

Reviewer 2: I think we don't need these chapters. Instead, make a prerequisites chapter to instruct on how to set up the Tuya IoT Platform, next include our config flow instructions.

You have created an account for [Tuya Smart or Smart Life app](https://developer.tuya.com/en/docs/iot/tuya-smart-app-smart-life-app-advantages?id=K989rqa49rluq#title-1-Download).

#### Create a project

1. Log in to the [Tuya IoT Platform](https://iot.tuya.com/?_source=627739af7d7e42379954f5c26a1c1759).
2. In the left navigation bar, click **Cloud** > **Projects**.
3. On the page that appears, click **Create**.
4. In the **Create Project** dialog box, configure **Project Name**, **Description**, **Industry**, and **Availability Zone**. In the **Development Method** field, select **Smart Home PaaS** from the dropdown list.
	> **Note:** Select all available zones.
	

> ![](../images/integrations/tuya/image_001.png)

Reviewer 2: Please add a description for each image, so screenreaders know what they are looking at.

```suggestion
![](/images/integrations/tuya/image_001.png)
```

	
5. Click **Create** to continue project configuration.
6. In **Configuration Wizard**, select **Device status notification** API.
   > **Note:** **Smart Home PaaS** API has been selected as default.
   

> ![](../images/integrations/tuya/image_002.png)

Reviewer 2: Don't use relative paths, as they are not relative to this repository but should work on our website.

Please make sure to test if the documentation renders as expected

```suggestion
![](/images/integrations/tuya/image_002.png)
```

7. Click **Authorize**.

#### Get authorization key

Click the created project to enter the **Project Overview** page and get the **Authorization Key** used to make API calls.

> ![](../images/integrations/tuya/image_003.png)

Reviewer 2: broken image

#### Link devices by app account

Link devices by your app account and copy the **Device ID** in the **Device List** to the clipboard for specifying `device_id` later.

> **1.** Navigate to the **Devices** tab.

Reviewer 2: We don't need to add emphasis to all words. Please reduce the use of emphasis used.

**2.** Click **Link Tuya App Account** > **Add App Account**.

> ![](../images/integrations/tuya/image_004.png)

Reviewer 2: broken image

**3.** Scan the QR code that appears using the **Tuya Smart** app or **Smart Life** app.

> ![](../images/integrations/tuya/image_005.png)

Reviewer 2: broken image

**4.** Click **Confirm** on your **Tuya Smart** or **Smart Life** app.

**5.** Navigate to the **Device List** tab. You can find the devices that have been added through your **Tuya Smart** or **Smart Life** account.   

> ![](../images/integrations/tuya/image_006.png)

Reviewer 2: broken image

</details>

### Install the Tuya Home Assistant integration

> #### Install Home Assistant

Reviewer 2: This is not the Home Assistant installation manual and doesn't need to be part of our documentation. Please remove this section.

Please refer to the [Home Assistant Official Installation](https://www.home-assistant.io/installation/) documentation to install **Home Assistant Core**.

You can also get help from [Set up Home Assistant Development Environment on Raspberry Pi](https://developer.tuya.com/en/demo/setuphomeassistantdevenv/?_source=b1441bb02314be3e594a0448891aac58) in Tuya Developer Demo Center.

> #### Install the Tuya integration

Reviewer 2: This whole section needs to be replaced with our config flow instructions and can be fully replace by:

`{% include integrations/config_flow.md %}`

In **Configuration** > **Integrations** > **ADD INTEGRATION**, search for the keyword **Tuya** and select it for installation.
![](../images/integrations/tuya/image_007.png)

#### Enter your Tuya credential

In the Tuya Integration window, select **Smart Home PaaS** and click **SUBMIT**.

> ![](../images/integrations/tuya/image_008.png)

Reviewer 2: Broken image

See the following table and enter your Tuya credential.

| Field | Description |
| ------- | -------- |
| Access ID and Access Secret| Go to your cloud project on [Tuya IoT Platform](https://iot.tuya.com/?_source=627739af7d7e42379954f5c26a1c1759). Find the **Access ID** and **Access Secret** on the **Project Overview** tab.|
| Mobile App | Must select the one you used to link devices on the Tuya IoT Platform. |
| Country Code | The country you select on logging in to Tuya Smart or Smart Life app.|
| Account | Tuya Smart or Smart Life app account. |
| Password | The password of your app account. |
> **Note**: The app mentioned in the table must be the one you used to link devices on the Tuya IoT Platform.
</details>

## Error codes and troubleshooting

|Error Code|Error Message|Troubleshooting|
|:----|:--------|:------------------------------|

> |1004| sign invalid| Incorrect Access ID or Access Secret. Please refer to [Tuya credential table](https://github.com/tuya/tuya-home-assistant/wiki/Install-Tuya-v2?_source=d10de34623e3daca5b02e3c31528a0c4#3-enter-your-tuya-credential).

Reviewer 2: ```suggestion
|1004| sign invalid| Incorrect Access ID or Access Secret. Please refer to [Tuya credential table](https://github.com/tuya/tuya-home-assistant/wiki/Install-Tuya-v2#3-enter-your-tuya-credential).
```

> |1100|param is empty| Empty parameter of username or app. Fill the parameters refer to [Tuya credential table](https://github.com/tuya/tuya-home-assistant/wiki/Install-Tuya-v2?_source=d10de34623e3daca5b02e3c31528a0c4#3-enter-your-tuya-credential).

Reviewer 2: ```suggestion
|1100|param is empty| Empty parameter of username or app. Fill the parameters refer to [Tuya credential table](https://github.com/tuya/tuya-home-assistant/wiki/Install-Tuya-v2?#3-enter-your-tuya-credential).
```

| 28841105 |No permissions. This project is not authorized to call this API| Some APIs are not authorized, please  [Subscribe](https://developer.tuya.com/en/docs/iot/applying-for-api-group-permissions?id=Ka6vf012u6q76#title-2-Subscribe%20to%20cloud%20products) then [Authorize](https://developer.tuya.com/en/docs/iot/applying-for-api-group-permissions?id=Ka6vf012u6q76#title-3-Authorize%20projects%20to%20call%20the%20cloud%20product). The following APIs must be subscribed for this tutorial: <ul><li>Authorization</li><li>Smart Home Devices Management</li><li>Smart Home Family Management</li><li>Smart Home Scene Linkage</li><li>Smart Home Data Service</li><li>Device status notification</li></ul>|

## Related Documents

> ### How to Contribute

Reviewer 2: Personally, I don't think this section belongs in the end-user documentation and should be removed.

```suggestion
### How to Contribute

```

Author: fixed

> * [How to Develop a New Driver](https://github.com/tuya/tuya-home-assistant/wiki/How-to-Develop-a-New-Driver)
> * [How to Get the Log](https://github.com/tuya/tuya-home-assistant/wiki/How-to-Get-the-Log)
> * [Supported Device Categories](https://github.com/tuya/tuya-home-assistant/wiki/Supported-Device-Category)

Reviewer 2: ```suggestion
- [How to Develop a New Driver](https://github.com/tuya/tuya-home-assistant/wiki/How-to-Develop-a-New-Driver)
- [How to Get the Log](https://github.com/tuya/tuya-home-assistant/wiki/How-to-Get-the-Log)
- [Supported Device Categories](https://github.com/tuya/tuya-home-assistant/wiki/Supported-Device-Category)
```

Author: fixed.

### Documentation

> * [How does the Integration work](https://github.com/tuya/tuya-home-assistant/wiki/How-the-Integration-works)

Reviewer 2: Should this be part of the documentation here?

```suggestion
- [How does the Integration work](https://github.com/tuya/tuya-home-assistant/wiki/How-the-Integration-works)
```

Author: fixed

> * [Set up Home Assistant on a Raspberry Pi](https://github.com/tuya/tuya-home-assistant/wiki/Set-up-Home-Assistant-on-a-Raspberry-Pi)

Reviewer 2: It is weird to link to an external website for this... lets remove it.

```suggestion
```

Author: fixed.

> * [FAQs](https://github.com/tuya/tuya-home-assistant/wiki/FAQs)

Reviewer 2: Shouldn't the FAQ be part of the documentation here?

```suggestion
- [FAQs](https://github.com/tuya/tuya-home-assistant/wiki/FAQs)
```

Author: fixed.

The Tuya Home Assistant integration is developed for controlling **Powered by Tuya (PBT)** devices, which is officially maintained by the Tuya Developer Team.

> You can find more information for this integration on [Tuya Integration Github](https://github.com/tuya/tuya-home-assistant?_source=418498f8d2819b5b1654a0c12c5325cd).

Reviewer 2: This is the documentation for the core integration, we should not refer to the external integration in the core documentation.

```suggestion
```

Author: fixed.

#### Create a project

> 1. Log in to the [Tuya IoT Platform](https://iot.tuya.com/?_source=627739af7d7e42379954f5c26a1c1759).

Reviewer 2: We value privacy and should not add any form of tracking on links.

```suggestion
1. Log in to the [Tuya IoT Platform](https://iot.tuya.com/).
```

Author: fixed.

| Field | Description |
| ------- | -------- |

> | Access ID and Access Secret| Go to your cloud project on [Tuya IoT Platform](https://iot.tuya.com/?_source=627739af7d7e42379954f5c26a1c1759). Find the **Access ID** and **Access Secret** on the **Project Overview** tab.|

Reviewer 2: ```suggestion
| Access ID and Access Secret| Go to your cloud project on [Tuya IoT Platform](https://iot.tuya.com/). Find the **Access ID** and **Access Secret** on the **Project Overview** tab.|
```

Author: fixed

|Error Code|Error Message|Troubleshooting|
|:----|:--------|:------------------------------|
|1004| sign invalid| Incorrect Access ID or Access Secret. Please refer to [Tuya credential table](https://github.com/tuya/tuya-home-assistant/wiki/Install-Tuya-v2?_source=d10de34623e3daca5b02e3c31528a0c4#3-enter-your-tuya-credential).

> |1106|permission deny|<ul><li> App account not linked with cloud project: On the [Tuya IoT Platform](https://iot.tuya.com/cloud/?_source=fd9f339c3f58e01cdc0b275d5bd97a21), you have linked devices by using Tuya Smart or Smart Life app in your cloud project. For more information, see [Link devices by app account](https://developer.tuya.com/en/docs/iot/Platform_Configuration_smarthome?id=Kamcgamwoevrx&_source=7a356dd493196a01bb9021b7680a2a45#title-3-Link%20devices%20by%20app%20account).</li><li> Incorrect username or password: Enter the correct account and password of the Tuya Smart or Smart Life app in the **Account** and **Password** fields. Note that the app account depends on which app (Tuya Smart or Smart Life) you used to link devices on the [Tuya IoT Platform](https://iot.tuya.com/cloud/?_source=fd9f339c3f58e01cdc0b275d5bd97a21).</li><li>Incorrect availability zone: See [Availability Zone](https://github.com/tuya/tuya-home-assistant/wiki/Tuya-IoT-Platform-Configuration-Guide-Using-Smart-Home-PaaS#region--available-zone-correspondence) and select the correct availability zone.</li><li>Incorrect countycode: Fill the [code](https://countrycode.org/) of the country you select on logging in to Tuya Smart or Smart Life app.</li></ul>|

Reviewer 2: ```suggestion
|1106|permission deny|<ul><li> App account not linked with cloud project: On the [Tuya IoT Platform](https://iot.tuya.com/cloud/), you have linked devices by using Tuya Smart or Smart Life app in your cloud project. For more information, see [Link devices by app account](https://developer.tuya.com/en/docs/iot/Platform_Configuration_smarthome?id=Kamcgamwoevrx#title-3-Link%20devices%20by%20app%20account).</li><li> Incorrect username or password: Enter the correct account and password of the Tuya Smart or Smart Life app in the **Account** and **Password** fields. Note that the app account depends on which app (Tuya Smart or Smart Life) you used to link devices on the [Tuya IoT Platform](https://iot.tuya.com/cloud/).</li><li>Incorrect availability zone: See [Availability Zone](https://github.com/tuya/tuya-home-assistant/wiki/Tuya-IoT-Platform-Configuration-Guide-Using-Smart-Home-PaaS#region--available-zone-correspondence) and select the correct availability zone.</li><li>Incorrect country code: Fill the [code](https://countrycode.org/) of the country you select on logging in to Tuya Smart or Smart Life app.</li></ul>|
```

Author: fixed

|1100|param is empty| Empty parameter of username or app. Fill the parameters refer to [Tuya credential table](https://github.com/tuya/tuya-home-assistant/wiki/Install-Tuya-v2?_source=d10de34623e3daca5b02e3c31528a0c4#3-enter-your-tuya-credential).
|2017|schema does not exist| Incorrect app selected. Select the app you used to link devices in the cloud project.|

> | 2406 | skill id invalid | Make sure that your cloud project on the [Tuya IoT Platform](https://iot.tuya.com/cloud/?_source=fd9f339c3f58e01cdc0b275d5bd97a21) should be created after May 25, 2021. Otherwise, you need to create a new project or migrate data to a new project. For more information, see [Operation on the Tuya IoT Platform](https://developer.tuya.com/en/docs/iot/migrate-from-an-older-version?id=Kamee9wtbd00b#title-3-Operation%20on%20the%20Tuya%20IoT%20Platform).|

Reviewer 2: ```suggestion
| 2406 | skill id invalid | Make sure that your cloud project on the [Tuya IoT Platform](https://iot.tuya.com/cloud/) should be created after May 25, 2021. Otherwise, you need to create a new project or migrate data to a new project. For more information, see [Operation on the Tuya IoT Platform](https://developer.tuya.com/en/docs/iot/migrate-from-an-older-version?id=Kamee9wtbd00b#title-3-Operation%20on%20the%20Tuya%20IoT%20Platform).|
```

Reviewer 2: ```suggestion
| 2406 | skill id invalid | Make sure that your cloud project on the [Tuya IoT Platform](https://iot.tuya.com/cloud/) should be created after May 25, 2021. Otherwise, you need to create a new project or migrate data to a new project. For more information, see [Operation on the Tuya IoT Platform](https://developer.tuya.com/en/docs/iot/migrate-from-an-older-version?id=Kamee9wtbd00b#title-3-Operation%20on%20the%20Tuya%20IoT%20Platform).|
```

Author: fixed
